File: scoring/propagator/reference/test_weighted_shortest_path.py
"""Unit tests for the weighted-shortest-path reference implementation.

Mirrors the canonical scenarios from
``tests/test_weighted_shortest_path.cc`` on the *same fixed inputs* so
the C++ and Python engines have a shared fixed-point. The parity
property test (Phase 4.4 + ADR finalization) generalizes this with
hypothesis-generated graphs.
"""

from __future__ import annotations

import math

import pytest

from scoring.propagator.reference.types import Edge, GraphData, Params
from scoring.propagator.reference.weighted_shortest_path import (
    name,
    propagate,
    version,
)

_TOLERANCE = 1e-9


def _unnormalized(max_d: int, alpha: float) -> Params:
    return Params(k_hop_radius=max_d, decay_weight=alpha, normalize=False)


def test_name_and_version() -> None:
    """The reference engine reports the same name and semver as the C++ engine."""
    assert name() == "weighted-shortest-path"
    assert version() == "0.1.0"


def test_trivial_one_node_graph_has_zero_uplift() -> None:
    graph = GraphData(node_ids=(42,), adjacency=((),), inputs=(7.0,))
    uplift = propagate(graph, _unnormalized(max_d=2, alpha=0.5))
    assert math.isclose(uplift[42], 0.0, abs_tol=_TOLERANCE)


def test_linear_chain_unit_weights() -> None:
    """3-node directed chain 0->1->2 with unit weights, alpha=1.0, max_d=2.

    Expected (matches test_weighted_shortest_path.cc):
      uplift[0] = 0
      uplift[1] = exp(-1)
      uplift[2] = exp(-2) + exp(-1)
    """
    graph = GraphData(
        node_ids=(0, 1, 2),
        adjacency=(
            (Edge(target=1, weight=1.0),),
            (Edge(target=2, weight=1.0),),
            (),
        ),
        inputs=(1.0, 1.0, 1.0),
    )
    uplift = propagate(graph, _unnormalized(max_d=2, alpha=1.0))
    assert math.isclose(uplift[0], 0.0, abs_tol=_TOLERANCE)
    assert math.isclose(uplift[1], math.exp(-1.0), abs_tol=_TOLERANCE)
    assert math.isclose(uplift[2], math.exp(-2.0) + math.exp(-1.0), abs_tol=_TOLERANCE)


def test_distance_cutoff_excludes_farther_targets() -> None:
    """A target beyond max_distance receives zero uplift from that source."""
    graph = GraphData(
        node_ids=(0, 1, 2),
        adjacency=(
            (Edge(target=1, weight=1.5),),
            (Edge(target=2, weight=1.5),),
            (),
        ),
        inputs=(1.0, 0.0, 0.0),
    )
    uplift = propagate(graph, _unnormalized(max_d=2, alpha=1.0))
    assert math.isclose(uplift[0], 0.0, abs_tol=_TOLERANCE)
    assert math.isclose(uplift[1], math.exp(-1.5), abs_tol=_TOLERANCE)
    assert math.isclose(uplift[2], 0.0, abs_tol=_TOLERANCE)


def test_zero_input_source_contributes_nothing() -> None:
    """No source has positive input -> zero uplift everywhere."""
    graph = GraphData(
        node_ids=(0, 1, 2),
        adjacency=(
            (Edge(target=1, weight=1.0),),
            (Edge(target=2, weight=1.0),),
            (),
        ),
        inputs=(0.0, 0.0, 0.0),
    )
    uplift = propagate(graph, _unnormalized(max_d=5, alpha=1.0))
    assert all(math.isclose(v, 0.0, abs_tol=_TOLERANCE) for v in uplift.values())


def test_disconnected_components_are_isolated() -> None:
    """Two directed 2-node components with different input scales stay isolated."""
    graph = GraphData(
        node_ids=(100, 101, 200, 201),
        adjacency=(
            (Edge(target=1, weight=1.0),),
            (),
            (Edge(target=3, weight=1.0),),
            (),
        ),
        inputs=(1.0, 0.0, 100.0, 0.0),
    )
    uplift = propagate(graph, _unnormalized(max_d=5, alpha=1.0))
    assert math.isclose(uplift[100], 0.0, abs_tol=_TOLERANCE)
    assert math.isclose(uplift[101], math.exp(-1.0), abs_tol=_TOLERANCE)
    assert math.isclose(uplift[200], 0.0, abs_tol=_TOLERANCE)
    assert math.isclose(uplift[201], 100.0 * math.exp(-1.0), abs_tol=_TOLERANCE)


def test_normalize_scales_by_maximum() -> None:
    """With normalize=true the max uplift equals 1.0 and ratios are preserved."""
    graph = GraphData(
        node_ids=(0, 1, 2),
        adjacency=(
            (Edge(target=1, weight=1.0),),
            (Edge(target=2, weight=1.0),),
            (),
        ),
        inputs=(1.0, 1.0, 1.0),
    )
    uplift = propagate(graph, Params(k_hop_radius=2, decay_weight=1.0, normalize=True))
    expected_1 = math.exp(-1.0) / (math.exp(-2.0) + math.exp(-1.0))
    assert math.isclose(uplift[0], 0.0, abs_tol=_TOLERANCE)
    assert math.isclose(uplift[1], expected_1, abs_tol=_TOLERANCE)
    assert math.isclose(uplift[2], 1.0, abs_tol=_TOLERANCE)


def test_invalid_graph_raises() -> None:
    """A graph that fails is_valid() must raise ValueError."""
    bad_graph = GraphData(
        node_ids=(0, 1),
        adjacency=((Edge(target=99, weight=1.0),), ()),
        inputs=(1.0, 1.0),
    )
    with pytest.raises(ValueError, match="invariant"):
        propagate(bad_graph, Params())

File: scoring/propagator/tests/test_weighted_shortest_path.cc
// WeightedShortestPath strategy unit tests — Phase 4.8.
//
// Pins the algorithm contract on canonical graphs that exercise the
// Dijkstra-driven decay semantics. Tests use normalize=false so the
// expected uplift values are direct functions of exp(-alpha * d).

#include <cmath>
#include <memory>
#include <string>
#include <vector>

#include "gtest/gtest.h"

#include "streetsense/propagator/graph.h"
#include "streetsense/propagator/registry.h"
#include "streetsense/propagator/strategy.h"

namespace {

using streetsense::propagator::Edge;
using streetsense::propagator::GraphData;
using streetsense::propagator::Params;
using streetsense::propagator::PropagationStrategy;
using streetsense::propagator::StrategyRegistry;

constexpr double kTolerance = 1e-9;

std::unique_ptr<PropagationStrategy> make_strategy() {
    auto strategy = StrategyRegistry::instance().lookup("weighted-shortest-path");
    EXPECT_NE(strategy, nullptr)
        << "weighted-shortest-path must be statically registered";
    return strategy;
}

Params unnormalized_params(int max_d, double alpha) {
    Params p;
    p.k_hop_radius = max_d;
    p.decay_weight = alpha;
    p.normalize = false;
    return p;
}

TEST(WeightedShortestPathTest, StrategyIsRegistered) {
    const auto names = StrategyRegistry::instance().list_names();
    bool found = false;
    for (const auto& name : names) {
        if (name == "weighted-shortest-path") {
            found = true;
            break;
        }
    }
    EXPECT_TRUE(found);
}

TEST(WeightedShortestPathTest, NameAndVersion) {
    auto strategy = make_strategy();
    ASSERT_NE(strategy, nullptr);
    EXPECT_EQ(strategy->name(), "weighted-shortest-path");
    EXPECT_EQ(strategy->version(), "0.1.0");
}

TEST(WeightedShortestPathTest, TrivialOneNodeGraphHasZeroUplift) {
    GraphData graph;
    graph.node_ids = {42};
    graph.adjacency = {{}};
    graph.inputs = {7.0};

    auto strategy = make_strategy();
    const auto uplift = strategy->propagate(graph, unnormalized_params(2, 0.5));

    ASSERT_EQ(uplift.count(42), 1U);
    EXPECT_NEAR(uplift.at(42), 0.0, kTolerance);
}

TEST(WeightedShortestPathTest, LinearChainUnitWeights) {
    // 3-node directed chain 0->1->2 with unit edge weights. inputs all
    // 1.0. alpha=1.0, max_d=2. Expected uplifts (only forward-reachable):
    //   uplift[1] = inputs[0]*exp(-1*1) + (inputs[2] doesn't reach 1) = e^-1
    //   uplift[2] = inputs[0]*exp(-1*2) + inputs[1]*exp(-1*1) = e^-2 + e^-1
    //   uplift[0] = 0 (no incoming edges, no shorter path from later sources)
    GraphData graph;
    graph.node_ids = {0, 1, 2};
    graph.adjacency = {
        {Edge{1, 1.0}},
        {Edge{2, 1.0}},
        {},
    };
    graph.inputs = {1.0, 1.0, 1.0};

    auto strategy = make_strategy();
    const auto uplift = strategy->propagate(graph, unnormalized_params(2, 1.0));

    EXPECT_NEAR(uplift.at(0), 0.0, kTolerance);
    EXPECT_NEAR(uplift.at(1), std::exp(-1.0), kTolerance);
    EXPECT_NEAR(uplift.at(2), std::exp(-2.0) + std::exp(-1.0), kTolerance);
}

TEST(WeightedShortestPathTest, DistanceCutoffExcludesFartherTargets) {
    // 0 -> 1 (w=1.5) -> 2 (w=1.5). With max_distance=2.0, source 0
    // reaches 1 (d=1.5) but not 2 (d=3.0 > 2.0).
    GraphData graph;
    graph.node_ids = {0, 1, 2};
    graph.adjacency = {
        {Edge{1, 1.5}},
        {Edge{2, 1.5}},
        {},
    };
    graph.inputs = {1.0, 0.0, 0.0};

    auto strategy = make_strategy();
    const auto uplift = strategy->propagate(graph, unnormalized_params(2, 1.0));

    EXPECT_NEAR(uplift.at(0), 0.0, kTolerance);
    EXPECT_NEAR(uplift.at(1), std::exp(-1.5), kTolerance);
    EXPECT_NEAR(uplift.at(2), 0.0, kTolerance);
}

TEST(WeightedShortestPathTest, ZeroInputSourceContributesNothing) {
    // Same chain but all inputs are zero -> zero uplift everywhere.
    GraphData graph;
    graph.node_ids = {0, 1, 2};
    graph.adjacency = {
        {Edge{1, 1.0}},
        {Edge{2, 1.0}},
        {},
    };
    graph.inputs = {0.0, 0.0, 0.0};

    auto strategy = make_strategy();
    const auto uplift = strategy->propagate(graph, unnormalized_params(5, 1.0));
    for (const auto& [node_id, value] : uplift) {
        static_cast<void>(node_id);
        EXPECT_NEAR(value, 0.0, kTolerance);
    }
}

TEST(WeightedShortestPathTest, DisconnectedComponentsAreIsolated) {
    // Component 1: 0 -> 1 with unit weight. Component 2: 2 -> 3 with
    // unit weight. With large max_distance, C1 nodes only see C1
    // contributions and vice versa.
    GraphData graph;
    graph.node_ids = {100, 101, 200, 201};
    graph.adjacency = {
        {Edge{1, 1.0}},
        {},
        {Edge{3, 1.0}},
        {},
    };
    graph.inputs = {1.0, 0.0, 100.0, 0.0};

    auto strategy = make_strategy();
    const auto uplift = strategy->propagate(graph, unnormalized_params(5, 1.0));

    EXPECT_NEAR(uplift.at(100), 0.0, kTolerance);
    EXPECT_NEAR(uplift.at(101), std::exp(-1.0), kTolerance);
    EXPECT_NEAR(uplift.at(200), 0.0, kTolerance);
    EXPECT_NEAR(uplift.at(201), 100.0 * std::exp(-1.0), kTolerance);
}

TEST(WeightedShortestPathTest, NormalizeScalesByMaximum) {
    // Reuse the linear chain (forward-directed) where the max uplift
    // is at node 2. With normalize=true, max(uplift)=1.0 and ratios
    // are preserved.
    GraphData graph;
    graph.node_ids = {0, 1, 2};
    graph.adjacency = {
        {Edge{1, 1.0}},
        {Edge{2, 1.0}},
        {},
    };
    graph.inputs = {1.0, 1.0, 1.0};

    Params params;
    params.k_hop_radius = 2;
    params.decay_weight = 1.0;
    params.normalize = true;

    auto strategy = make_strategy();
    const auto uplift = strategy->propagate(graph, params);

    // Pre-normalization: uplift[0]=0, uplift[1]=e^-1≈0.3679,
    // uplift[2]=e^-2 + e^-1 ≈ 0.5032. Max = uplift[2].
    // Post-normalization: uplift[1] = (e^-1)/(e^-2 + e^-1) = e/(1+e).
    const double expected_1 = std::exp(-1.0) / (std::exp(-2.0) + std::exp(-1.0));
    EXPECT_NEAR(uplift.at(0), 0.0, kTolerance);
    EXPECT_NEAR(uplift.at(1), expected_1, kTolerance);
    EXPECT_NEAR(uplift.at(2), 1.0, kTolerance);
}

}  // namespace
